src/components/statevector_component/aggregation.py

```diff
@@ -0,0 +1,390 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# SBATCH -N 1
+# SBATCH -n 1
+
+import xarray as xr
+import numpy as np
+import yaml
+import copy
+import sys
+import time
+from inversion_scripts.imi_preview import estimate_averaging_kernel
+
+# clustering
+from sklearn.cluster import KMeans
+
+
+def match_data_to_clusters(data, clusters, default_value=0):
+    """
+    Description:
+        match 1D list of elements to 2D statevector
+    arguments:
+        data                [int]: data to infill into 2D statevector
+        clusters            [][] : xarray dataarray: the mapping from
+                                   the grid cells to state vector number
+    Returns:                [][] : filled with new values
+    """
+    result = clusters.copy()
+    c_array = result.values
+    c_idx = np.where(c_array > 0)
+    c_val = c_array[c_idx]
+    row_idx = [r for _, r, _ in sorted(zip(c_val, c_idx[0], c_idx[1]))]
+    col_idx = [c for _, _, c in sorted(zip(c_val, c_idx[0], c_idx[1]))]
+    idx = (row_idx, col_idx)
+
+    d_idx = np.where(c_array == 0)
+
+    c_array[c_idx] = data
+    c_array[d_idx] = default_value
+    result.values = c_array
+
+    return result
+
+
+def zero_buffer_elements(clusters, num_buffer_elems):
+    """
+    Description:
+        Return clusters with buffer elements set to 0 and
+        buffers with roi set to 0
+    arguments:
+        clusters            [][] : xarray dataarray: the mapping
+                                  from the grid cells to state vector number
+        num_buffer_elems    int  : number of buffer elements in the state vector
+    Returns:                tuple: zeroed buffer area and zeroed roi
+    """
+    buffer_threshold = int(clusters.max()) - num_buffer_elems
+    result = (clusters <= buffer_threshold) * clusters
+    buffers = (clusters > buffer_threshold) * clusters
+    return result, buffers
+
+
+def scale_buffer_elements(scale_number, buffer):
```

Comment:
Not sure I understand what this is doing. Shouldn't the buffer elements be static while the ROI elements get clustered together?

Reply:
The buffer elements need to be relabeled once there are fewer sv elements. For example, if there are originally 63 elements in the DOI and 8 buffer elements labeled 64-71. Once the DOI elements have been clustered into, say, 47 elements, the buffer elements need to be relabeled to 48-55. Maybe I should add an additional comment to the description